### PR TITLE
binserve: init at 0.2.0

### PR DESCRIPTION
**Pull request by snapdgn**
###### Description of changes
Add binserve: a http server in Rust

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/binserve/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  src = fetchFromGitHub {
+    owner = "mufeedvh";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-Chm2xPB0BrLXSZslg9wnbDyHSJRQAvOtpH0Rw6w1q1s=";
+  };
+
+  cargoLock.lockFile = ./Cargo.lock;
+  postPatch = "cp ${./Cargo.lock} Cargo.lock";
```
Comment (azahi) [inline pkgs/servers/binserve/default.nix:15]:
```suggestion
  postPatch = ''
    cp ${./Cargo.lock} Cargo.lock
  '';
```

**Review by azahi — CHANGES_REQUESTED** [review]

File: pkgs/servers/binserve/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    cp ${./Cargo.lock} Cargo.lock
+  '';
+
+  meta = with lib; {
+    description = "A fast production-ready static web server";
+    homepage = "https://github.com/mufeedvh/binserve";
+    longDescription = ''
+    A fast production-ready static web server with TLS
+    (HTTPS), routing, hot reloading, caching, templating, and security in a
+    single-binary you can set up with zero code
```
Comment (viraptor) [inline pkgs/servers/binserve/default.nix:25]: nixpkgs-fmt says this should be:
```suggestion
      A fast production-ready static web server with TLS
      (HTTPS), routing, hot reloading, caching, templating, and security in a
      single-binary you can set up with zero code
```

File: pkgs/servers/binserve/default.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{ lib, stdenv, fetchFromGitHub, rustPlatform }:
```
Comment (viraptor) [inline pkgs/servers/binserve/default.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub, rustPlatform, CoreServices }:
```

File: pkgs/servers/binserve/default.nix [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  pname = "binserve";
+  version = "0.2.0";
+
+  src = fetchFromGitHub {
+    owner = "mufeedvh";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-Chm2xPB0BrLXSZslg9wnbDyHSJRQAvOtpH0Rw6w1q1s=";
+  };
+
```
Comment (viraptor) [inline pkgs/servers/binserve/default.nix:13]:
```suggestion

  buildInputs = lib.optionals stdenv.isDarwin [ CoreServices ];

```

File: pkgs/servers/binserve/default.nix [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    description = "A fast production-ready static web server";
+    homepage = "https://github.com/mufeedvh/binserve";
+    longDescription = ''
+    A fast production-ready static web server with TLS
+    (HTTPS), routing, hot reloading, caching, templating, and security in a
+    single-binary you can set up with zero code
+    '';
+    license = licenses.mit;
+    maintainers = with maintainers; [ snapdgn ];
+    platforms = platforms.linux;
```
Comment (viraptor) [inline pkgs/servers/binserve/default.nix:29]:
```suggestion
    platforms = platforms.unix;
```

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -275,6 +275,8 @@ with pkgs;
 
   bingo = callPackage ../development/tools/bingo {};
 
+  binserve = callPackage ../servers/binserve { };
```
Comment (viraptor) [inline pkgs/top-level/all-packages.nix:278]:
```suggestion
  binserve = callPackage ../servers/binserve {
    inherit (darwin.apple_sdk.frameworks) CoreServices;
  };
```

**Review by viraptor — COMMENTED** [review]
This works on non-linux platforms too. With the suggested fixes, it's verified working on intel darwin.

File: pkgs/servers/binserve/default.nix [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-Chm2xPB0BrLXSZslg9wnbDyHSJRQAvOtpH0Rw6w1q1s=";
+  };
+
+  cargoLock.lockFile = ./Cargo.lock;
+  postPatch = ''
+    cp ${./Cargo.lock} Cargo.lock
+  '';
+
```
Comment (viraptor) [inline pkgs/servers/binserve/default.nix:18]: I don't believe there are any tests in the source code, so it just builds twice without a good reason.
```suggestion

  doCheck = false;

```

**Comment by snapdgn** [comment]
Thanks @viraptor , I've incorporated the changes, please take a look again!

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -275,6 +275,10 @@ with pkgs;
 
   bingo = callPackage ../development/tools/bingo {};
 
+  binserve = callPackage ../servers/binserve { 
+    inherit (darwin.apple-sdk.frameworks) CoreServices;
```
Comment (viraptor) [inline pkgs/top-level/all-packages.nix:279]: There's a typo:
```suggestion
    inherit (darwin.apple_sdk.frameworks) CoreServices;
```

Comment (snapdgn) [reply pkgs/top-level/all-packages.nix:279]: yeah, sorry about that :P, fixed now.

**Comment by viraptor** [comment]
Result of `nixpkgs-review pr 179364` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>binserve</li>
  </ul>
</details>

Seems to work fine.

Please check the editorconfig results (https://github.com/NixOS/nixpkgs/runs/7087250661?check_suite_fo cus=true) and make sure you squash your PR into two commits (one for adding the maintainer entry and one for adding the package).
Otherwise 👍

**Comment by snapdgn** [comment]
> Result of `nixpkgs-review pr 179364` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> 
> 1 package built:
> Seems to work fine.
> 
> Please check the editorconfig results (https://github.com/NixOS/nixpkgs/runs/7087250661?check_suite_focus=true) and make sure you squash your PR into two commits (one for adding the maintainer entry and one for adding the package). Otherwise +1

**Review by viraptor — APPROVED** [review]

File: maintainers/maintainer-list.nix [hunk 9]
```diff
@@ -14725,4 +14725,10 @@
     github = "npatsakula";
     githubId = 23001619;
   };
+  snapdgn = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:14728]: This entry must be sorted alphabetical.

File: pkgs/servers/binserve/default.nix [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    sha256 = "sha256-Chm2xPB0BrLXSZslg9wnbDyHSJRQAvOtpH0Rw6w1q1s=";
+  };
+
+  buildInputs = lib.optionals stdenv.isDarwin [
+    CoreServices
+  ];
+
+  cargoLock.lockFile = ./Cargo.lock;
+  postPatch = ''
+    cp ${./Cargo.lock} Cargo.lock
```
Comment (SuperSandro2000) [inline pkgs/servers/binserve/default.nix:20]: Can't we use fetchCrate to not vendor the lock file?

File: pkgs/servers/binserve/default.nix [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    description = "A fast production-ready static web server";
+    homepage = "https://github.com/mufeedvh/binserve";
+    longDescription = ''
+      A fast production-ready static web server with TLS
+      (HTTPS), routing, hot reloading, caching, templating, and security in a
+      single-binary you can set up with zero code
+    '';
+    license = licenses.mit;
+    maintainers = with maintainers; [ snapdgn ];
+    platforms = platforms.unix;
```
Comment (SuperSandro2000) [inline pkgs/servers/binserve/default.nix:35]: That is the default platform for rust and shouldn't be repeated.

File: pkgs/servers/binserve/default.nix (hunk 10 above)
Comment (snapdgn) [reply pkgs/servers/binserve/default.nix:20]: The package is not present on crates.io, so i guess we can't go along with that.

**Pull request merged**